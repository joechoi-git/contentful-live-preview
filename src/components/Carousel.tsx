import React from "react";
import Image from "next/image";
import type { BynderAsset } from "../lib/contentful/adjustedTypes";
import { transformBynderAsset } from "@/lib/contentful/Utils";

interface CarouselProps {
    slides: BynderAsset[];
    width: number;
    height: number;
}

const Carousel: React.FC<CarouselProps> = ({ slides, width, height }) => {
    return slides && slides.length > 0 ? (
        <>
            <div className="carousel w-full p-4">
                {slides?.map((slide, index) => {
                    return (
                        <div
                            id={`carousel-${index}`}
                            key={`${index}-${slide.id}`}
                            className="carousel-item w-full mr-2"
                        >
                            <Image
                                alt={
                                    slide.description && slide.description.length > 0
                                        ? slide.description
                                        : slide.name
                                }
                                height={height}
                                width={width}
                                unoptimized={true}
                                src={transformBynderAsset(
                                    slide,
                                    `io=transform:extend,width:${width},height:${height},background:ffff00`
                                )}
                                className="rounded-xl"
                            />
                        </div>
                    );
                })}
            </div>
            <div className="flex justify-center w-full py-2 gap-2">
                {slides?.map((slide, index) => {
                    return (
                        <a
                            href={`#carousel-${index}`}
                            key={`${index}-${slide.id}`}
                            className="btn bg-secondary hover:bg-accent text-secondary-content"
                        >
                            {index++}
                        </a>
                    );
                })}
            </div>
        </>
    ) : null;
};

export default Carousel;
