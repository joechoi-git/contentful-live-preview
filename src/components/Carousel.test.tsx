import React from "react";
import { render, screen } from "@testing-library/react";
import Carousel from "./Carousel";
import type { BynderAsset } from "../lib/contentful/adjustedTypes";

const mockSlides: BynderAsset[] = [
    {
        archive: 0,
        brandId: "brand1",
        copyright: null,
        dateCreated: "2023-01-01",
        dateModified: "2023-01-02",
        datePublished: "2023-01-03",
        description: "Slide 1 description",
        extension: ["jpg"],
        fileSize: 12345,
        height: 400,
        id: "slide1",
        isPublic: 1,
        limited: 0,
        name: "Slide 1",
        orientation: "landscape",
        original: "https://example.com/slide1.jpg",
        thumbnails: {
            webimage: "https://example.com/slide1_webimage.jpg",
            thul: "https://example.com/slide1_thul.jpg",
            original: "https://example.com/slide1_original.jpg",
            mini: "https://example.com/slide1_mini.jpg",
            transformBaseUrl: "https://example.com/slide1_transform.jpg",
            Social: "https://example.com/slide1_social.jpg"
        },
        type: "image",
        watermarked: 0,
        width: 1200,
        videoPreviewURLs: [],
        tags: [],
        textMetaproperties: [],
        src: "https://example.com/slide1.jpg"
    },
    {
        archive: 0,
        brandId: "brand2",
        copyright: null,
        dateCreated: "2023-02-01",
        dateModified: "2023-02-02",
        datePublished: "2023-02-03",
        description: "Slide 2 description",
        extension: ["jpg"],
        fileSize: 12346,
        height: 400,
        id: "slide2",
        isPublic: 1,
        limited: 0,
        name: "Slide 2",
        orientation: "landscape",
        original: "https://example.com/slide2.jpg",
        thumbnails: {
            webimage: "https://example.com/slide2_webimage.jpg",
            thul: "https://example.com/slide2_thul.jpg",
            original: "https://example.com/slide2_original.jpg",
            mini: "https://example.com/slide2_mini.jpg",
            transformBaseUrl: "https://example.com/slide2_transform.jpg",
            Social: "https://example.com/slide2_social.jpg"
        },
        type: "image",
        watermarked: 0,
        width: 1200,
        videoPreviewURLs: [],
        tags: [],
        textMetaproperties: [],
        src: "https://example.com/slide2.jpg"
    }
];

describe("Carousel", () => {
    it("renders without crashing", () => {
        render(<Carousel slides={mockSlides} width={1200} height={400} />);
        expect(screen.getByAltText("Slide 1 description")).toBeInTheDocument();
        expect(screen.getByAltText("Slide 2 description")).toBeInTheDocument();
    });

    it("displays the correct number of slides", () => {
        render(<Carousel slides={mockSlides} width={1200} height={400} />);
        const slides = screen.getAllByRole("img");
        expect(slides).toHaveLength(2);
    });

    it("displays navigation buttons", () => {
        render(<Carousel slides={mockSlides} width={1200} height={400} />);
        const buttons = screen.getAllByRole("link");
        expect(buttons).toHaveLength(2);
        expect(buttons[0]).toHaveAttribute("href", "#carousel-0");
        expect(buttons[1]).toHaveAttribute("href", "#carousel-1");
    });

    it("matches snapshot", () => {
        const { asFragment } = render(<Carousel slides={mockSlides} width={1200} height={400} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
