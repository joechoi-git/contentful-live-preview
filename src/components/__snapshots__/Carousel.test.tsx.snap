// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="carousel w-full p-4"
  >
    <div
      class="carousel-item w-full mr-2"
      id="carousel-0"
    >
      <img
        alt="Slide 1 description"
        class="rounded-xl"
        data-nimg="1"
        decoding="async"
        height="400"
        loading="lazy"
        src="https://example.com/slide-1?format=webp&date=2024-07-25-17-02&quality=80&io=transform:extend,width:1200,height:400,background:cccccc"
        style="color: transparent;"
        width="1200"
      />
    </div>
    <div
      class="carousel-item w-full mr-2"
      id="carousel-1"
    >
      <img
        alt="Slide 2 description"
        class="rounded-xl"
        data-nimg="1"
        decoding="async"
        height="400"
        loading="lazy"
        src="https://example.com/slide-2?format=webp&date=2024-07-25-17-02&quality=80&io=transform:extend,width:1200,height:400,background:cccccc"
        style="color: transparent;"
        width="1200"
      />
    </div>
  </div>
  <div
    class="flex justify-center w-full py-2 gap-2"
  >
    <a
      class="btn bg-secondary hover:bg-accent text-secondary-content"
      href="#carousel-0"
    >
      0
    </a>
    <a
      class="btn bg-secondary hover:bg-accent text-secondary-content"
      href="#carousel-1"
    >
      1
    </a>
  </div>
</DocumentFragment>
`;
